### DISCUSSION TITLE
Generate RDT Token for the restricted APIs

### DISCUSSION THREAD

**Discussion by hardikradixweb** [amazon-php/sp-api-sdk#202] [Q&A] [answered]
Hello,

I have implemented pull order functionality using this SDK.

Here's what I have implemented, I am able to retrieve orders and order items but I am not able to fetch buyer information and shipping address, for that, I have to replace Access Token with Restricted Data Token but I am not able to implement RDT in this process can you please Help/Guide me that how I can implement RDT in this! and also is this appropriate way to integrate SDK like this? if not then can you please guide me so I can update accordingly?

`  
          // Set region
           $region = Regions::NORTH_AMERICA;
            if ($apiData['body']['Region'] == 'us-west-2') {
                $region = Regions::FAR_EAST;
            } else if ($apiData['body']['Region'] == 'eu-west-1') {
                $region = Regions::EUROPE;
            }
            $factory = new Psr17Factory();
            $client = new Curl($factory);
           
            $sts = new STSClient(
                $client,
                $requestFactory = $factory,
                $streamFactory = $factory
            );
           
            $logger = new Logger('name');
            $logger->pushHandler(new StreamHandler(storage_path('logs/sp-api-php.log'), Logger::DEBUG));
            
            $oauth = new OAuth(
                $client,
                $httpFactory = new HttpFactory($requestFactory, $streamFactory),
                $config = Configuration::forIAMRole(
                    $apiData['body']['AppID'],
                    $apiData['body']['AppSecret'],
                    $sts->assumeRole(
                        $apiData['body']['DeveloperAccessKeyID'],
                        $apiData['body']['DeveloperSecretAccessKey'],
                        $apiData['body']['RoleARN']
                    )
                ),
                $logger
            );
            $sdk = SellingPartnerSDK::create($client, $factory, $factory, $config, $logger);
            $accessToken = $oauth->exchangeRefreshToken($apiData['body']['AppRefreshToken']);
            
            try {
                $orderResponse = $sdk->orders()->getOrders(
                    $accessToken,
                    $region,
                    $marketplaceId = ['ATVPDKIKX0DER,A1AM78C64UM0Y8,A1MQXOICRS2Z7M,A2EUQ1WTGCTBG2,A2ZV50J4W1RKNI,A3H6HPSLHAK3XG,AHRY1CZE9ZY4H'],
                    $created_after = Carbon::now()->subDays(1)->toIso8601ZuluString('millisecond')
                );
                $ordersData = $orderResponse->getPayload()->getOrders();
                $orderArr['orders'] = $ordersData;
                if (!empty($orderArr['orders'])) {
                    foreach ($orderArr['orders'] as $orderKey => $order) {
                        $orderItems = $sdk->orders()->getOrderItems(
                            $accessToken,
                            $region,
                            $order['amazon_order_id']
                        );
                        $ordersItemsData = $orderItems->getPayload()->getOrderItems();
                        $orderArr['orders'][$orderKey]['orderItems'] = $ordersItemsData;
                        sleep(2);
                    }
                }
                $filteredData = $this->filterOrderData($orderArr);
                return array('status' => 'success', 'data' => $filteredData);
            } catch (ApiException $exception) {
                return array('status' => 'error', 'data' => $exception->getMessage());
            }`

**Comment by norberttech** [comment]
hey, could you please start from providing the error message you are getting from Amazon?

**Reply by hardikradixweb** [reply]
I am not getting any errors from Amazon.

I am able to retrieve all the data via PostMan calls, and in my current integration I am able to fetch Orders and Order Items but not the Buyer Information as it is restricted so for that I have to send Restricted Data Token but I did not found any way to pass Restricted Data Token in the SDK.

I have to pass RDT instead of Access Token in the getOrders

**Reply by norberttech** [reply]
Have you checked the [official documentation](https://developer-docs.amazon.com/sp-api/docs/regulated-orders#tutorial-how-to-retrieve-an-orders-regulated-information)?

**Reply by hardikradixweb** [reply]
Yes, and I do have access to those restricted data

**Reply by hardikradixweb** [reply]
I am stuck with the SDK that how can I implement that in my code

**Reply by hardikradixweb** [reply]
and the code snipped that I have attached it is working fine, and now I want to integrate RDT in that code so I can fetch order buyer information together

**Reply by norberttech** [reply]
Is your developer and application properly authorized? Since RDT is not available if your profile is not verified: 

Please find requirements below: 

```
* Approval for the Direct-to-Consumer Delivery (Restricted) role in your developer profile.
* The Direct-to-Consumer Delivery (Restricted) role selected in the App registration page for your application.
```

**Reply by hardikradixweb** [reply]
Yes, it is properly authorised and i am able to retrieve buyer information in the postman calls

**Reply by hardikradixweb** [reply]
Do you have any example that can explain usage of RTD in this SDK?

**Reply by hardikradixweb** [reply]
Hello @norberttech,

Would you like to connect? I am
having an issue with passing RDT in this SDK and provided code snippet, it would be great if you can help to resolve this. 

![Amazon Order Response](https://user-images.githubusercontent.com/152806/176639408-9ac14eda-75ed-4f36-9f8f-7ae4d1f474a6.png)

**Comment by norberttech** [comment, marked as answer]
Please find a working example of accessing buyerInfo and shippingAddress below: 

```php
<?php

use AmazonPHP\SellingPartner\AccessToken;
use AmazonPHP\SellingPartner\Model\Tokens\RestrictedResource;
use AmazonPHP\SellingPartner\Regions;
use AmazonPHP\SellingPartner\SellingPartnerSDK;
use Buzz\Client\Curl;
use AmazonPHP\SellingPartner\Configuration;
use Monolog\Handler\StreamHandler;
use Monolog\Logger;
use Nyholm\Psr7\Factory\Psr17Factory;
use AmazonPHP\SellingPartner\Model\Tokens\CreateRestrictedDataTokenRequest;

require_once __DIR__ . '/vendor/autoload.php';

$factory = new Psr17Factory();
$client = new Curl($factory);

$sellerId = 'xxx';
$sellerRefreshToken = 'xxx';

$clientId = 'xxx';
$clientSecret = 'xxx';
$accessKey = 'xxx';
$secretKey = 'xx';

$orderId = 'xxx-xxxxxxx-xxxxxxx';

$configuration = Configuration::forIAMUser(
    $clientId,
    $clientSecret,
    $accessKey,
    $secretKey
);

$logger = new Logger('name');
$logger->pushHandler(new StreamHandler(__DIR__ . '/sp-api-php.log', Logger::DEBUG));

$sdk = SellingPartnerSDK::create($client, $factory, $factory, $configuration, $logger);

$accessToken = $sdk->oAuth()->exchangeRefreshToken($sellerRefreshToken);

$restrictedDataAccessToken = $sdk->tokens()->createRestrictedDataToken(
    $accessToken,
    Regions::NORTH_AMERICA,
    new CreateRestrictedDataTokenRequest([
        'restricted_resources' => [
            new RestrictedResource([
                'method' => RestrictedResource::METHOD_GET,
                'path' => '/orders/v0/orders/{orderId}',
                'data_elements' => ['buyerInfo', 'shippingAddress'],
            ]),
        ],
    ])
);

$order = $sdk->orders()->getOrder(
    new AccessToken(
        $restrictedDataAccessToken->getRestrictedDataToken(),
        'not_available',
        'bearer',
        $restrictedDataAccessToken->getExpiresIn(),
        'restricted_data_token'
    ),
    Regions::NORTH_AMERICA,
    $orderId
);

dump($order);

```

That's the best I can do, I'm not doing any consulting over social media or any other communication channels, this library is generated from [amazon models](https://github.com/amzn/selling-partner-api-models) repository so it should work as long as you are following [docs](https://github.com/amzn/selling-partner-api-docs/)

**Reply by hardikradixweb** [reply]
I really appreciate your effort! A big thank you, I will try to implement this.

**Reply by chrism-delmarfans** [reply]
I was all over the place trying to find out what I needed to replace refresh token with. Thank you for this example!

**Reply by nhtahoe** [reply]
Thank you! Amazon did not make this obvious at all, but your example made it simple to implement calls to Orders and Reports APIs and get PII in return.

**Comment by xiaoguo0426** [comment]
![image](https://user-images.githubusercontent.com/8944962/209087150-7b5745f4-c77a-401b-868c-71c365395a9f.png)
@norberttech    I got the above error according to your code.
